feat: add docs for how to deploy the filesystem client (#17)
charmed-hpc/docs: howto/getting-started/deploy-shared-filesystem.md

(deploy-shared-filesystem)=
# How to deploy a shared filesystem

Charmed-HPC allows automatic integration with shared filesystems using the
[filesystem-client](https://charmhub.io/filesystem-client) charm. This how-to guide shows you how to
deploy `filesystem-client` to integrate with externally managed shared filesystems.

## Prerequisites

- {ref}`deploy-slurm`.

## Deploy an external filesystem server

External servers that provide a shared filesystem cannot be integrated directly. Instead,
we can use a [proxy charm](https://juju.is/docs/sdk/charm-taxonomy#heading--proxy) in order to expose
the required information to applications managed by Juju.

:::::::{tabs}

::::::{group-tab} NFS

To integrate with an external NFS server, you will require:
- An externally managed NFS server.
- The server's hostname.
- The exported path.
- (optional) the port.

> Each public cloud has its own procedure to deploy a public NFS server. Provided here are links to
> every procedure on some of the most well-known public clouds.

Reviewer: ```suggestion
Each public cloud has its own procedure to deploy a public NFS server. Provided here are links to
the set up procedures on a few well-known public clouds.
```

- Amazon Web Services 
    - See [Create an NFS file share using the default configuration][aws-nfs] for general set up information.
- Azure
    - See [Tutorial: Create an NFS Azure file share and mount it on a Linux VM using the Azure portal][azure-nfs]

>       for general set up information.

Reviewer: This could probably be formatted more cleanly but I'm not quite sure what that would look like - I'll think about it

Author: Made some adjustments to use clickable cards instead of a bullet list. Let me know if this looks better!

Reviewer: I like the cards! I might see if we can tweak the background color so they stand out a bit more (darn dark mode), but they're functional as is.

(deploy-shared-filesystem)=
# How to deploy a shared filesystem

Charmed-HPC allows automatic integration with shared filesystems using the
[filesystem-client](https://charmhub.io/filesystem-client) charm. This how-to guide shows you how to
deploy `filesystem-client` to integrate with externally managed shared filesystems.

## Prerequisites

- {ref}`deploy-workload-manager`.

## Deploy an external filesystem server

Any external servers providing a shared filesystem cannot be directly integrated with. In this case,
we can use a [proxy charm](https://juju.is/docs/sdk/charm-taxonomy#heading--proxy) in order to expose
the required information to applications managed by Juju.

:::::::{tabs}

::::::{group-tab} NFS

To integrate with an external NFS server, you will require:
- An externally managed NFS server.
- The server's hostname.
- The exported path.
- (optional) the port.

Each public cloud has it's own procedure to deploy a public NFS server. We provide some
links on how to deploy a NFS server in the most well-known public clouds.

However, if you only want a minimal server for testing purposes, we also explain how to setup a

> small NFS server on LXD.

Author: Probably not needed anymore since the tabs were removed?

:::::{tabs}

::::{group-tab} Amazon Web Services

See ["Create an NFS file share using the default configuration"][aws-nfs] for more information.

[aws-nfs]: https://docs.aws.amazon.com/filegateway/latest/files3/nfs-fileshare-quickstart-settings.html

::::

::::{group-tab} Azure

See ["Tutorial: Create an NFS Azure file share and mount it on a Linux VM using the Azure portal"][azure-nfs]
for more information.

[azure-nfs]: https://learn.microsoft.com/en-us/azure/storage/files/storage-files-quick-create-use-linux

> 

Author: Removed the tabs and added individual sections for the public clouds.

    --config auth-info=fs-client:<value of $CLIENT_KEY>
:::

::::::

:::::::

## Deploy the filesystem-client

To add the charm that takes care of mounting a shared filesystem to the cluster nodes,
deploy the `filesystem-client` charm:

:::{code-block} shell
juju deploy filesystem-client --channel latest/edge \
    --config mountpoint='/scratch' \
    --config noexec=true
:::

The `mountpoint` configuration represents the path that the filesystem will be mounted onto.

`filesystem-client` is a [subordinate charm](https://juju.is/docs/juju/relation#heading--subordinate)
that can automatically mount any shared filesystems for the application related with it.
In this case, we will relate it to the `slurmd` application in order to have a shared storage between
all the compute nodes in the cluster:

:::{code-block} shell
juju integrate slurmd:juju-info filesystem-client:juju-info
:::

## Relate the filesystem client with the filesystem provider

Every filesystem provider can be integrated with the filesystem client using the
`filesystem` endpoint.

:::{code-block} shell
juju integrate filesystem-client:filesystem <filesystem-provider>:filesystem
:::

> Afterwards, the filesystem should be accessible to read and write from the `slurmd` application

Reviewer: ```suggestion
Afterwards, test that the filesystem is accessible to read and write from the `slurmd` application
```
